### PR TITLE
Fixes #3563: Sha3.pm

### LINKED ISSUE #3563: Remove `html_enc` from Goodies. No longer needed :)
for fixing bug #3563

### PR DESCRIPTION
**Pull request by gautamkrishnar**
Fixes #3563: Sha3.pm removed html_enc()
cc: @moollaza 

---
Instant Answer Page: https://duck.co/ia/view/sha3

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @rafacas

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by gautamkrishnar** [comment]
@moollaza why the tests are failing?

File: t/SHA3.t [hunk 1]
```diff
@@ -191,12 +191,12 @@ ddg_goodie_test(
...
         'b8d157e78f7b40565da31d547bb269ae08784d44bdf508759583fdc8c4aacddef53f45af7f272444d040e92b286501c60337665ae1a1e0d4326049b3a5d6ae5e',
-        '&lt;script&gt;alert(&quot;ddg&quot;)&lt;/script&gt;',
+        '<script>alert("ddg")</script>',
         'SHA3-512 hex hash'
     ),
-    'sha3-512 & / " \\\' ; < >' => build_test(
+    'sha3-512 \& / " \\\' ; < >' => build_test(
         '5fd58e6c60e027e8fc237098bac75c5fbddd33ee4a3b95868fb902ce555ae6b873783ee08fdfd90fbf4232d00ac755a0abd0d6fa1a25b277589846dfdba0c64b',
-        '&amp; / &quot; \&#39; ; &lt; &gt;',
+        '\& / " \&#39; ; < >',
```
Comment (mintsoft) [inline t/SHA3.t]: @gautamkrishnar you need to substitute `\&#39`  ; this is why the tests are failing

Comment (gautamkrishnar) [inline t/SHA3.t]: Thanks :smile:

File: t/SHA3.t [hunk 2]
```diff
@@ -75,8 +75,8 @@ ddg_goodie_test(
         'SHA3-224 hex hash'
     ),
     'sha3-224 \& / " \\\' ; < >' => build_test(
```
Comment (mintsoft) [inline t/SHA3.t]: Ifyou change this to be:
```perl
q|sha3-224 \& / " ' ; < >| => build_test(
``` 
Then the hash should be the same as it was before the changes. If it isn't then we need to investigate why it's changed

Comment (gautamkrishnar) [inline t/SHA3.t]: @mintsoft  Yes the hash is changing when i am using the ` '\& / " \\\' ; < >'`. i am not sure why it is changing...
the test is only failing for the queries with a `'` in it. that's why i replaced the hash.

Comment (gautamkrishnar) [inline t/SHA3.t]: @mintsoft i am still receiving a new hash and test is failing even after using `q|sha3-224 \& / " ' ; < >| => build_test(`:
screenshot 119

See my  code:
screenshot 120

Comment (mintsoft) [inline t/SHA3.t]: @gautamkrishnar You've left `\'` in; so perl will be treating that as a literal `\'` I think if you change it to just `'` then it'll work?

Comment (gautamkrishnar) [inline t/SHA3.t]: @mintsoft still my tests are failing... 
```
[01:12 PM codio@harvest-ozone zeroclickinfo-goodies {patch-10}]$ duckpan test sha3                                                                                                            
t/SHA3.t .. 1/?                                                                                                                                                                               
    #   Failed test 'Deep: full ZCI object'                                                                                                                                                   
    #   at /home/codio/perl5/perls/perl-5.18.2/lib/site_perl/5.18.2/DDG/Test/Goodie.pm line 54.                                                                                               
    # Compared $data->{"answer"}                                                                                                                                                              
    #    got : 'e11c539bc9f76958d0810986ef889effb1b59fda8020f576fbd6d430'                                                                                                                     
    # expect : '7c5deb70175f99dd6b1edcd0b6e75b34be47f9fb385c48f8612e3e7c'                                                                                                                     
    # Looks like you failed 1 test of 1.                                                                                                                                                      
                                                                                                                                                                                              
#   Failed test 'Query: sha3-224 \& / " \' ; < >'                                                                                                                                             
#   at /home/codio/perl5/perls/perl-5.18.2/lib/site_perl/5.18.2/DDG/Test/Goodie.pm line 58.                                                                                                   
# Looks like you failed 1 test of 50.                                                                                                                                                         
t/SHA3.t .. Dubious, test returned 1 (wstat 256, 0x100)                                                                                                                                       
Failed 1/50 subtests                                                                                                                                                                          
                                                                                                                                                                                              
Test Summary Report                                                                                                                                                                           
-------------------                                                                                                                                                                           
t/SHA3.t (Wstat: 256 Tests: 50 Failed: 1)                                                                                                                                                     
  Failed test:  11                                                                                                                                                                            
  Non-zero exit status: 1                                                                                                                                                                     
Files=1, Tests=50,  0 wallclock secs ( 0.04 usr  0.00 sys +  0.23 cusr  0.02 csys =  0.29 CPU)                                                                                                
Result: FAIL                                                                                                                                                                                  
[ERROR]  Tests failed! See output above for details                 
```

i used the code below:
```
q|sha3-224 \& / " \\' ; < >| => build_test(
        '7c5deb70175f99dd6b1edcd0b6e75b34be47f9fb385c48f8612e3e7c',
        '\& / " \\\' ; < >',
        'SHA3-224 hex hash'
    )
```
also tried using:
```
q|sha3-224 & / " \\' ; < >| => build_test(
        '7c5deb70175f99dd6b1edcd0b6e75b34be47f9fb385c48f8612e3e7c',
        '\& / " \\\' ; < >',
        'SHA3-224 hex hash'
    )
```
but it is throwing this error on the test
```:
[01:13 PM codio@harvest-ozone zeroclickinfo-goodies {patch-10}]$ duckpan test sha3                                                                                                            
t/SHA3.t .. 1/?                                                                                                                                                                               
    #   Failed test 'Deep: full ZCI object'                                                                                                                                                   
    #   at /home/codio/perl5/perls/perl-5.18.2/lib/site_perl/5.18.2/DDG/Test/Goodie.pm line 54.                                                                                               
    # Compared $data->{"structured_answer"}{"data"}{"subtitle"}                                                                                                                               
    #    got : 'SHA3-224 hex hash: & / " \' ; < >'                                                                                                                                            
    # expect : 'SHA3-224 hex hash: \& / " \' ; < >'                                                                                                                                           
    # Looks like you failed 1 test of 1.                                                                                                                                                      
                                                                                                                                                                                              
#   Failed test 'Query: sha3-224 & / " \' ; < >'                                                                                                                                              
#   at /home/codio/perl5/perls/perl-5.18.2/lib/site_perl/5.18.2/DDG/Test/Goodie.pm line 58.                                                                                                   
# Looks like you failed 1 test of 50.                                                                                                                                                         
t/SHA3.t .. Dubious, test returned 1 (wstat 256, 0x100)                                                                                                                                       
Failed 1/50 subtests                                                                                                                                                                          
                                                                                                                                                                                              
Test Summary Report                                                                                                                                                                           
-------------------                                                                                                                                                                           
t/SHA3.t (Wstat: 256 Tests: 50 Failed: 1)                                                                                                                                                     
  Failed test:  11                                                                                                                                                                            
  Non-zero exit status: 1                                                                                                                                                                     
Files=1, Tests=50,  1 wallclock secs ( 0.04 usr  0.00 sys +  0.22 cusr  0.02 csys =  0.28 CPU)                                                                                                
Result: FAIL                                                                                                                                                                                  
[ERROR]  Tests failed! See output above for details     
```

**Comment by GuiltyDolphin** [comment]
@gautamkrishnar Looks like Sha3 has been done already - thanks for the PR though!

**Comment by gautamkrishnar** [comment]
@GuiltyDolphin  okay thanks... i was actually struggling with this one,.,,,

**Comment by GuiltyDolphin** [comment]
@gautamkrishnar Sorry about this - looks like the entry was ticked before the Goodie was actually updated - meaning this PR can continue if you want.

If you want to continue, just pop over to your branches on GH and there should be a little trash-can icon next to the branch and you can un-delete it.

Sorry about that :frowning:

**Comment by gautamkrishnar** [comment]
@GuiltyDolphin undeleted :+1: 
Can you please help me with the tests?

File: lib/DDG/Goodie/SHA3.pm [hunk 3]
```diff
@@ -42,8 +42,8 @@ handle query => sub {
 
     return $out, structured_answer => {
         data => {
-            title => html_enc($out),
-            subtitle => html_enc(uc($alg) . "-$ver $enc hash").": ".html_enc($str)
+            title => $out,
+            subtitle => uc($alg) . "-$ver $enc hash".": ".$str
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/SHA3.pm]: `"-$ver $enc hash: $str"`

**Comment by GuiltyDolphin** [comment]
@gautamkrishnar The tests look pretty happy to me - thanks!

File: lib/DDG/Goodie/SHA3.pm (hunk 3 above)
Comment (gautamkrishnar) [inline lib/DDG/Goodie/SHA3.pm]: Thanks :smile:  :flags:

**Pull request merged**